File: tools/stress_test.js
#!/usr/bin/env node
// stress_test.js — simulates N participants polling the npoint backend
// and measures success rate, latency, and 429 rate.
//
// Usage: node stress_test.js <binId> <participants> <durationSeconds>
// Example: node stress_test.js abc123 10 30

const [,, BIN_ID, PARTICIPANTS_STR, DURATION_STR] = process.argv;
const PARTICIPANTS  = parseInt(PARTICIPANTS_STR || '10', 10);
const DURATION_MS  = parseInt(DURATION_STR || '30', 10) * 1000;
const POLL_INTERVAL_MS = 5000;  // matches app.js

if (!BIN_ID) {
  console.error('Usage: node stress_test.js <binId> <participants> <durationSeconds>');
  process.exit(1);
}

const URL = `https://api.npoint.io/${BIN_ID}`;

let stats = {
  reads:    { ok: 0, rate429: 0, err: 0, totalMs: 0, count: 0 },
  writes:   { ok: 0, rate429: 0, err: 0, totalMs: 0, count: 0 },
};

async function read(label) {
  const t0 = Date.now();
  try {
    const r = await fetch(URL);
    const ms = Date.now() - t0;
    stats.reads.count++;
    stats.reads.totalMs += ms;
    if (r.status === 429) { stats.reads.rate429++; return null; }
    if (!r.ok)            { stats.reads.err++;     return null; }
    stats.reads.ok++;
    return await r.json();
  } catch(e) {
    stats.reads.err++;
    return null;
  }
}

async function write(data) {
  const t0 = Date.now();
  try {
    const r = await fetch(URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    const ms = Date.now() - t0;
    stats.writes.count++;
    stats.writes.totalMs += ms;
    if (r.status === 429) { stats.writes.rate429++; return false; }
    if (!r.ok)            { stats.writes.err++;     return false; }
    stats.writes.ok++;
    return true;
  } catch(e) {
    stats.writes.err++;
    return false;
  }
}

// Simulate one participant: poll every POLL_INTERVAL_MS
async function participant(id, stopAt) {
  // Stagger start so they don't all fire simultaneously
  await new Promise(r => setTimeout(r, Math.random() * POLL_INTERVAL_MS));

  while (Date.now() < stopAt) {
    await read(`p${id}`);
    await new Promise(r => setTimeout(r, POLL_INTERVAL_MS));
  }
}

// Simulate a voting burst (everyone votes within ~10 seconds)
async function votingBurst(stopAt) {
  await new Promise(r => setTimeout(r, 8000)); // wait 8s then burst
  if (Date.now() >= stopAt) return;

  console.log(`\n[${ts()}] Simulating voting burst (${PARTICIPANTS} votes)…`);
  const voters = Array.from({ length: PARTICIPANTS }, (_, i) => i);
  // Votes arrive spread over 10s (realistic webinar behaviour)
  await Promise.all(voters.map(async (id) => {
    await new Promise(r => setTimeout(r, Math.random() * 10000));
    if (Date.now() >= stopAt) return;
    const current = await read(`voter${id}`);
    if (current !== null) {
      const updated = { ...current, [`vote_p${id}`]: Math.floor(Math.random() * 4) };
      await write(updated);
    }
  }));
  console.log(`[${ts()}] Voting burst complete.`);
}

function ts() {
  return new Date().toISOString().slice(11, 19);
}

function printStats() {
  const elapsed = DURATION_MS / 1000;
  const r = stats.reads, w = stats.writes;
  const totalReqs = r.count + w.count;

  console.log('\n══════════════════════════════════════════');
  console.log(`  Stress test: ${PARTICIPANTS} participants × ${elapsed}s`);
  console.log('══════════════════════════════════════════');
  console.log(`  Total requests:  ${totalReqs}  (${(totalReqs / elapsed).toFixed(1)}/s, ${(totalReqs / elapsed * 60).toFixed(0)}/min)`);
  console.log('');
  console.log(`  READS  (${r.count} total)`);
  console.log(`    ✓ OK:       ${r.ok}   (${pct(r.ok, r.count)})`);
  console.log(`    ✗ 429:      ${r.rate429}  (${pct(r.rate429, r.count)})`);
  console.log(`    ✗ other:    ${r.err}`);
  console.log(`    avg latency: ${r.count ? Math.round(r.totalMs / r.count) : '—'}ms`);
  console.log('');
  console.log(`  WRITES (${w.count} total)`);
  console.log(`    ✓ OK:       ${w.ok}   (${pct(w.ok, w.count)})`);
  console.log(`    ✗ 429:      ${w.rate429}  (${pct(w.rate429, w.count)})`);
  console.log(`    ✗ other:    ${w.err}`);
  console.log(`    avg latency: ${w.count ? Math.round(w.totalMs / w.count) : '—'}ms`);
  console.log('');

  const failRate = (r.rate429 + w.rate429 + r.err + w.err) / Math.max(totalReqs, 1);
  if (failRate === 0) {
    console.log('  VERDICT: ✅  No failures at this scale.');
  } else if (failRate < 0.05) {
    console.log(`  VERDICT: ⚠️   ${pct(failRate*totalReqs, totalReqs)} failure rate — borderline.`);
  } else {
    console.log(`  VERDICT: ❌  ${pct(failRate*totalReqs, totalReqs)} failure rate — WILL NOT hold up.`);
  }
  console.log('══════════════════════════════════════════\n');
}

function pct(n, d) {
  if (!d) return '—';
  return (n / d * 100).toFixed(0) + '%';
}

// ── main ──────────────────────────────────────────────────────────────────────

const stopAt = Date.now() + DURATION_MS;

console.log(`\n[${ts()}] Starting stress test`);
console.log(`  Participants: ${PARTICIPANTS}`);
console.log(`  Duration:     ${DURATION_MS / 1000}s`);
console.log(`  Poll interval: ${POLL_INTERVAL_MS}ms`);
console.log(`  Bin: ${URL}\n`);

// Print a live counter every 5s
const ticker = setInterval(() => {
  const r = stats.reads, w = stats.writes;
  console.log(`[${ts()}]  reads: ${r.ok}ok/${r.rate429}×429  writes: ${w.ok}ok/${w.rate429}×429`);
}, 5000);

Promise.all([
  ...Array.from({ length: PARTICIPANTS }, (_, i) => participant(i, stopAt)),
  votingBurst(stopAt),
]).then(() => {
  clearInterval(ticker);
  printStats();
});

File: tools/stress_test.py
#!/usr/bin/env python3
"""
stress_test.py — simulates N participants polling the npoint backend
and measures success rate, latency, and 429 rate.

Usage: python3 stress_test.py <binId> <participants> <durationSeconds>
"""
import sys, asyncio, time, json, random
from urllib.request import urlopen, Request
from urllib.error import HTTPError, URLError
import threading

BIN_ID      = sys.argv[1] if len(sys.argv) > 1 else None
PARTICIPANTS = int(sys.argv[2]) if len(sys.argv) > 2 else 10
DURATION    = int(sys.argv[3]) if len(sys.argv) > 3 else 30
POLL_INTERVAL = 5.0  # seconds — matches app.js

if not BIN_ID:
    print("Usage: python3 stress_test.py <binId> <participants> <durationSeconds>")
    sys.exit(1)

URL = f"https://api.npoint.io/{BIN_ID}"

# Thread-safe stats
import threading
lock = threading.Lock()
stats = {
    "reads":  {"ok": 0, "r429": 0, "err": 0, "ms": 0, "n": 0},
    "writes": {"ok": 0, "r429": 0, "err": 0, "ms": 0, "n": 0},
}

def do_read():
    t0 = time.time()
    try:
        req = Request(URL, headers={"User-Agent": "stress-test"})
        with urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read())
            ms = int((time.time() - t0) * 1000)
            with lock:
                s = stats["reads"]
                s["ok"] += 1; s["ms"] += ms; s["n"] += 1
            return data
    except HTTPError as e:
        ms = int((time.time() - t0) * 1000)
        with lock:
            s = stats["reads"]
            s["n"] += 1; s["ms"] += ms
            if e.code == 429: s["r429"] += 1
            else:             s["err"] += 1
        return None
    except Exception:
        with lock:
            s = stats["reads"]
            s["err"] += 1; s["n"] += 1
        return None

def do_write(data):
    t0 = time.time()
    try:
        body = json.dumps(data).encode()
        req = Request(URL, data=body, method="POST",
                      headers={"Content-Type": "application/json",
                               "User-Agent": "stress-test"})
        with urlopen(req, timeout=10) as resp:
            resp.read()
            ms = int((time.time() - t0) * 1000)
            with lock:
                s = stats["writes"]
                s["ok"] += 1; s["ms"] += ms; s["n"] += 1
            return True
    except HTTPError as e:
        ms = int((time.time() - t0) * 1000)
        with lock:
            s = stats["writes"]
            s["n"] += 1; s["ms"] += ms
            if e.code == 429: s["r429"] += 1
            else:             s["err"] += 1
        return False
    except Exception:
        with lock:
            s = stats["writes"]
            s["err"] += 1; s["n"] += 1
        return False

def participant(pid, stop_at):
    # Stagger starts
    time.sleep(random.random() * POLL_INTERVAL)
    while time.time() < stop_at:
        do_read()
        time.sleep(POLL_INTERVAL)

def voting_burst(stop_at):
    """Everyone votes within ~10s of the 8s mark."""
    time.sleep(8)
    if time.time() >= stop_at:
        return
    print(f"\n[{ts()}] Simulating voting burst ({PARTICIPANTS} votes)…")

    def one_vote(vid):
        time.sleep(random.random() * 10)
        if time.time() >= stop_at:
            return
        current = do_read()
        if current is not None:
            current[f"vote_p{vid}"] = random.randint(0, 3)
            do_write(current)

    threads = [threading.Thread(target=one_vote, args=(i,), daemon=True)
               for i in range(PARTICIPANTS)]
    for t in threads: t.start()
    for t in threads: t.join()
    print(f"[{ts()}] Voting burst complete.")

def ts():
    return time.strftime("%H:%M:%S")

def print_stats():
    r, w = stats["reads"], stats["writes"]
    total = r["n"] + w["n"]
    rate_s  = total / DURATION
    rate_m  = rate_s * 60

    def pct(n, d):
        return f"{n/d*100:.0f}%" if d else "—"

    fail_n = r["r429"] + w["r429"] + r["err"] + w["err"]

    print("\n══════════════════════════════════════════")
    print(f"  Stress test: {PARTICIPANTS} participants × {DURATION}s")
    print(f"  (extrapolated to full 60-min webinar)")
    print("══════════════════════════════════════════")
    print(f"  Total requests : {total}  ({rate_s:.1f}/s, {rate_m:.0f}/min)")
    print(f"  Extrapolated   : ~{rate_m*60:.0f} requests over a 60-min session")
    print()
    print(f"  READS  ({r['n']} total)")
    print(f"    ✓ OK      : {r['ok']}  ({pct(r['ok'], r['n'])})")
    print(f"    ✗ 429     : {r['r429']}  ({pct(r['r429'], r['n'])})")
    print(f"    ✗ other   : {r['err']}")
    print(f"    avg latency: {r['ms']//r['n'] if r['n'] else '—'}ms")
    print()
    print(f"  WRITES ({w['n']} total)")
    print(f"    ✓ OK      : {w['ok']}  ({pct(w['ok'], w['n'])})")
    print(f"    ✗ 429     : {w['r429']}  ({pct(w['r429'], w['n'])})")
    print(f"    ✗ other   : {w['err']}")
    print(f"    avg latency: {w['ms']//w['n'] if w['n'] else '—'}ms")
    print()
    fail_rate = fail_n / total if total else 0
    if fail_rate == 0:
        verdict = "✅  No failures at this scale."
    elif fail_rate < 0.05:
        verdict = f"⚠️   {pct(fail_n, total)} failure rate — borderline, risky for live webinar."
    else:
        verdict = f"❌  {pct(fail_n, total)} failure rate — WILL NOT hold up for 100 participants."
    print(f"  VERDICT: {verdict}")
    print("══════════════════════════════════════════\n")

# ── main ─────────────────────────────────────────────────────────────────────

stop_at = time.time() + DURATION

print(f"\n[{ts()}] Starting stress test")
print(f"  Participants : {PARTICIPANTS}")
print(f"  Duration     : {DURATION}s")
print(f"  Poll interval: {POLL_INTERVAL}s")
print(f"  URL          : {URL}\n")

# Live ticker
def ticker_fn():
    while time.time() < stop_at:
        time.sleep(5)
        r, w = stats["reads"], stats["writes"]
        print(f"[{ts()}]  reads: {r['ok']}ok/{r['r429']}×429  "
              f"writes: {w['ok']}ok/{w['r429']}×429")

ticker = threading.Thread(target=ticker_fn, daemon=True)
ticker.start()

threads = [threading.Thread(target=participant, args=(i, stop_at), daemon=True)
           for i in range(PARTICIPANTS)]
burst = threading.Thread(target=voting_burst, args=(stop_at,), daemon=True)

for t in threads: t.start()
burst.start()
for t in threads: t.join()
burst.join()

print_stats()
